File: src/main/java/vazkii/tinkerer/common/block/tile/TileRepairer.java
/**
 x * This class was created by <Vazkii>. It's distributed as
 * part of the ThaumicTinkerer Mod.
 *
 * ThaumicTinkerer is Open Source and distributed under a
 * Creative Commons Attribution-NonCommercial-ShareAlike 3.0 License
 * (http://creativecommons.org/licenses/by-nc-sa/3.0/deed.en_GB)
 *
 * ThaumicTinkerer is a Derivative Work on Thaumcraft 4.
 * Thaumcraft 4 (c) Azanor 2012
 * (http://www.minecraftforum.net/topic/1585216-)
 *
 * File Created @ [Nov 30, 2013, 5:39:09 PM (GMT)]
 */
package vazkii.tinkerer.common.block.tile;

import appeng.api.movable.IMovableTile;
import cpw.mods.fml.common.Loader;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.ISidedInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.play.server.S35PacketUpdateTileEntity;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.common.util.Constants;
import net.minecraftforge.common.util.ForgeDirection;
import thaumcraft.api.ThaumcraftApiHelper;
import thaumcraft.api.aspects.Aspect;
import thaumcraft.api.aspects.AspectList;
import thaumcraft.api.aspects.IAspectContainer;
import thaumcraft.api.aspects.IEssentiaTransport;
import vazkii.tinkerer.common.ThaumicTinkerer;
import vazkii.tinkerer.common.compat.TinkersConstructCompat;
import vazkii.tinkerer.common.core.handler.ConfigHandler;
import vazkii.tinkerer.common.lib.LibBlockNames;

import java.util.HashMap;
import java.util.Map;

/*import vazkii.tinkerer.common.compat.TinkersConstructCompat;*/

public class TileRepairer extends TileEntity implements ISidedInventory, IAspectContainer, IEssentiaTransport, IMovableTile {

	int dmgLastTick = 0;
	public int ticksExisted = 0;
	public boolean tookLastTick = true;
	private static final Map<Aspect, Integer> repairValues = new HashMap();

	static {
		repairValues.put(Aspect.TOOL, 8);
		repairValues.put(Aspect.CRAFT, 5);
		repairValues.put(Aspect.ORDER, 3);
	}

	ItemStack[] inventorySlots = new ItemStack[1];

	@Override
	public void updateEntity() {
		if (++ticksExisted % 10 == 0) {
			if (Loader.isModLoaded("TConstruct") && ConfigHandler.repairTConTools) {
				if (inventorySlots[0] != null) {
					if (TinkersConstructCompat.isTConstructTool(inventorySlots[0])) {
						int dmg = TinkersConstructCompat.getDamage(inventorySlots[0]);
						if (dmg > 0) {
							int essentia = drawEssentia();
							TinkersConstructCompat.fixDamage(inventorySlots[0], essentia);
							markDirty();
							if (dmgLastTick != 0 && dmgLastTick != dmg) {
								ThaumicTinkerer.tcProxy.sparkle((float) (xCoord + 0.25 + Math.random() / 2F), (float) (yCoord + 1 + Math.random() / 2F), (float) (zCoord + 0.25 + Math.random() / 2F), 0);
								tookLastTick = true;
							} else tookLastTick = false;
						} else tookLastTick = false;
						dmgLastTick = inventorySlots[0] == null ? 0 : TinkersConstructCompat.getDamage(inventorySlots[0]);
						return;
					}
				}
			}
			if (inventorySlots[0] != null && inventorySlots[0].getItemDamage() > 0) {
				int essentia = drawEssentia();
				int dmg = inventorySlots[0].getItemDamage();
				inventorySlots[0].setItemDamage(Math.max(0, dmg - essentia));
				markDirty();

				if (dmgLastTick != 0 && dmgLastTick != dmg) {
					ThaumicTinkerer.tcProxy.sparkle((float) (xCoord + 0.25 + Math.random() / 2F), (float) (yCoord + 1 + Math.random() / 2F), (float) (zCoord + 0.25 + Math.random() / 2F), 0);
					tookLastTick = true;
				} else tookLastTick = false;
			} else tookLastTick = false;

			dmgLastTick = inventorySlots[0] == null ? 0 : inventorySlots[0].getItemDamage();
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound par1NBTTagCompound) {
		super.readFromNBT(par1NBTTagCompound);

		readCustomNBT(par1NBTTagCompound);
	}

	@Override
	public void writeToNBT(NBTTagCompound par1NBTTagCompound) {
		super.writeToNBT(par1NBTTagCompound);

		writeCustomNBT(par1NBTTagCompound);
	}

	public void readCustomNBT(NBTTagCompound par1NBTTagCompound) {
		NBTTagList nbttaglist = par1NBTTagCompound.getTagList("Items", Constants.NBT.TAG_COMPOUND);
		inventorySlots = new ItemStack[1];

		if (nbttaglist.tagCount() > 0) {
			NBTTagCompound tagList = nbttaglist.getCompoundTagAt(0);
			inventorySlots[0] = ItemStack.loadItemStackFromNBT(tagList);
		}
	}

	public void writeCustomNBT(NBTTagCompound par1NBTTagCompound) {
		NBTTagList nbttaglist = new NBTTagList();
		if (inventorySlots[0] != null) {
			NBTTagCompound tagList = new NBTTagCompound();
			tagList.setByte("Slot", (byte) 0);
			inventorySlots[0].writeToNBT(tagList);
			nbttaglist.appendTag(tagList);
		}
		par1NBTTagCompound.setTag("Items", nbttaglist);
	}

	@Override
	public int getSizeInventory() {
		return inventorySlots.length;
	}

	@Override
	public ItemStack getStackInSlot(int i) {
		return inventorySlots[i];
	}

	@Override
	public ItemStack decrStackSize(int i, int j) {
		if (inventorySlots[i] != null) {
			ItemStack stackAt;

			if (inventorySlots[i].stackSize <= j) {
				stackAt = inventorySlots[i];
				inventorySlots[i] = null;
				return stackAt;
			} else {
				stackAt = inventorySlots[i].splitStack(j);

				if (inventorySlots[i].stackSize == 0)
					inventorySlots[i] = null;

				return stackAt;
			}
		}

		return null;
	}

	@Override
	public ItemStack getStackInSlotOnClosing(int i) {
		return getStackInSlot(i);
	}

	@Override
	public void setInventorySlotContents(int i, ItemStack itemstack) {
		inventorySlots[i] = itemstack;
	}

	@Override
	public String getInventoryName() {
		return LibBlockNames.REPAIRER;
	}

	@Override
	public boolean hasCustomInventoryName() {
		return false;
	}

	@Override
	public int getInventoryStackLimit() {
		return 1;
	}

	@Override
	public boolean isUseableByPlayer(EntityPlayer entityplayer) {
		return worldObj.getTileEntity(xCoord, yCoord, zCoord) == this && entityplayer.getDistanceSq(xCoord + 0.5D, yCoord + 0.5D, zCoord + 0.5D) <= 64;
	}

	@Override
	public void openInventory() {

	}

	@Override
	public void closeInventory() {

	}

	@Override
	public boolean isItemValidForSlot(int i, ItemStack itemstack) {
		return true;
	}

	@Override
	public S35PacketUpdateTileEntity getDescriptionPacket() {
		NBTTagCompound nbttagcompound = new NBTTagCompound();
		writeCustomNBT(nbttagcompound);
		return new S35PacketUpdateTileEntity(xCoord, yCoord, zCoord, -999, nbttagcompound);
	}

	@Override
	public void onDataPacket(NetworkManager manager, S35PacketUpdateTileEntity packet) {
		super.onDataPacket(manager, packet);
		readCustomNBT(packet.func_148857_g());
	}

	@Override
	public void markDirty() {
		super.markDirty();
		worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
	}

	@Override
	public int[] getAccessibleSlotsFromSide(int var1) {
		return new int[]{ 0 };
	}

	@Override
	public boolean canInsertItem(int i, ItemStack itemstack, int j) {
		if (Loader.isModLoaded("TConstruct") && ConfigHandler.repairTConTools) {
			if (TinkersConstructCompat.isTConstructTool(itemstack)) {
				return itemstack != null;
			}
		}
		return itemstack != null && itemstack.getItem().isRepairable();
	}

	@Override
	public boolean canExtractItem(int i, ItemStack itemstack, int j) {
		return true;
	}

	int drawEssentia() {
		ForgeDirection orientation = getOrientation();
		TileEntity te = ThaumcraftApiHelper.getConnectableTile(worldObj, xCoord, yCoord, zCoord, orientation);
		if (te != null) {
			IEssentiaTransport ic = (IEssentiaTransport) te;
			if (!ic.canOutputTo(orientation.getOpposite()))
				return 0;

			for (Aspect aspect : repairValues.keySet())
				if (ic.getSuctionType(orientation.getOpposite()) == aspect && ic.getSuctionAmount(orientation.getOpposite()) < getSuctionAmount(orientation) && ic.takeEssentia(aspect, 1, orientation.getOpposite()) == 1)
					return repairValues.get(aspect);
		}
		return 0;
	}

	ForgeDirection getOrientation() {
		return ForgeDirection.getOrientation(getBlockMetadata());
	}

	@Override
	public AspectList getAspects() {
		ItemStack stack = inventorySlots[0];
		if (stack == null)
			return null;
		if (Loader.isModLoaded("TConstruct") && ConfigHandler.repairTConTools) {
			if (TinkersConstructCompat.isTConstructTool(stack))
				return new AspectList().add(Aspect.ENTROPY, TinkersConstructCompat.getDamage(stack));
		}
		return new AspectList().add(Aspect.ENTROPY, stack.getItemDamage());
	}

	@Override
	public void setAspects(AspectList paramAspectList) {
	}

	@Override
	public boolean doesContainerAccept(Aspect paramAspect) {

		return false;
	}

	@Override
	public int addToContainer(Aspect paramAspect, int paramInt) {
		return 0;
	}

	@Override
	public boolean takeFromContainer(Aspect paramAspect, int paramInt) {
		return false;
	}

	@Override
	public boolean takeFromContainer(AspectList paramAspectList) {
		return false;
	}

	@Override
	public boolean doesContainerContainAmount(Aspect paramAspect, int paramInt) {
		return false;
	}

	@Override
	public boolean doesContainerContain(AspectList paramAspectList) {
		return false;
	}

	@Override
	public int containerContains(Aspect paramAspect) {
		return 0;
	}

	@Override
	public boolean isConnectable(ForgeDirection paramForgeDirection) {
		return paramForgeDirection == getOrientation();
	}

	@Override
	public boolean canInputFrom(ForgeDirection paramForgeDirection) {
		return false;
	}

	@Override
	public boolean canOutputTo(ForgeDirection paramForgeDirection) {
		return isConnectable(paramForgeDirection);
	}

	@Override
	public void setSuction(Aspect paramAspect, int paramInt) {
	}

	@Override
	public int takeEssentia(Aspect paramAspect, int paramInt, ForgeDirection direction) {
		return 0;
	}

	@Override
	public int getMinimumSuction() {
		return 0;
	}

	@Override
	public boolean renderExtendedTube() {
		return false;
	}

	@Override
	public int addEssentia(Aspect arg0, int arg1, ForgeDirection direction) {
		return 0;
	}

	@Override
	public int getEssentiaAmount(ForgeDirection arg0) {
		return 0;
	}

	@Override
	public Aspect getEssentiaType(ForgeDirection arg0) {
		return null;
	}

	@Override
	public int getSuctionAmount(ForgeDirection arg0) {
		return arg0 == getOrientation() ? 128 : 0;
	}

	@Override
	public Aspect getSuctionType(ForgeDirection arg0) {
		return arg0 == getOrientation() ? Aspect.TOOL : null;
	}

	@Override
	public boolean prepareToMove() {
		return true;
	}

	@Override
	public void doneMoving() {

	}

}

File: src/main/scala/com/nekokittygames/Thaumic/Tinkerer/common/tiles/TileRepairer.scala
package com.nekokittygames.Thaumic.Tinkerer.common.tiles

import codechicken.lib.inventory.InventoryNBT
import com.nekokittygames.Thaumic.Tinkerer.common.blocks.BlockRepairer
import com.nekokittygames.Thaumic.Tinkerer.common.libs.LibNames
import net.minecraft.block.BlockPistonBase
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.{ISidedInventory, IInventory}
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.play.INetHandlerPlayClient
import net.minecraft.network.{NetworkManager, Packet}
import net.minecraft.network.play.server.S35PacketUpdateTileEntity
import net.minecraft.tileentity.TileEntity
import net.minecraft.util._
import net.minecraft.world.World
import thaumcraft.api.aspects.{AspectList, IAspectContainer, Aspect, IEssentiaTransport}
import thaumcraft.api.wands.IWandable



/**
 * Created by fiona on 21/05/2015.
 */
class TileRepairer extends TileEntity with IWandable with IEssentiaTransport with IAspectContainer with ISidedInventory with ITickable{

  override def update(): Unit = ticksExisted=ticksExisted+1

  var inventory:ItemStack=null
  var ticksExisted=0
// Create a custom def to read the inventory, single item only.
  def readCustomNBT(compound: NBTTagCompound)={
    inventory=ItemStack.loadItemStackFromNBT(compound.getCompoundTag("inventory"))
  }
  // Create a custom def to write if the inventory is empty, single item only.
  def writeCustomNBT(compound: NBTTagCompound)={
    val comp=new NBTTagCompound
    if (inventory!=null)
      inventory.writeToNBT(comp)
    compound.setTag("inventory",comp)
  }
  override def writeToNBT(compound: NBTTagCompound): Unit = {
    super.writeToNBT(compound)
    writeCustomNBT(compound)

  }
  override def getDescriptionPacket: Packet[INetHandlerPlayClient] = {
    super.getDescriptionPacket
    val compound:NBTTagCompound=new NBTTagCompound
    writeCustomNBT(compound)
    new S35PacketUpdateTileEntity(this.pos,1,compound)
  }

  override def onDataPacket(net: NetworkManager, pkt: S35PacketUpdateTileEntity): Unit = {
    super.onDataPacket(net, pkt)
    readCustomNBT(pkt.getNbtCompound)
  }

  override def readFromNBT(compound: NBTTagCompound): Unit = {
    super.readFromNBT(compound)
    readCustomNBT(compound)

  }


  override def shouldRefresh(world: World, pos: BlockPos, oldState: IBlockState, newSate: IBlockState): Boolean = false

  override def onWandStoppedUsing(itemStack: ItemStack, world: World, entityPlayer: EntityPlayer, i: Int): Unit = {}

  override def onUsingWandTick(itemStack: ItemStack, entityPlayer: EntityPlayer, i: Int): Unit = {}

  override def onWandRightClick(world: World, itemStack: ItemStack, entityPlayer: EntityPlayer, blockPos: BlockPos, enumFacing: EnumFacing): Boolean = {
    if(entityPlayer.isSneaking){
      world.setBlockState(blockPos,world.getBlockState(blockPos).withProperty(BlockRepairer.FACING,BlockPistonBase.getFacingFromEntity(world,blockPos,entityPlayer)))
    }else{
      world.setBlockState(blockPos,world.getBlockState(blockPos).withProperty(BlockRepairer.FACING,BlockPistonBase.getFacingFromEntity(world,blockPos,entityPlayer).getOpposite))
    }
    this.markDirty()
    true
  }

  override def addEssentia(aspect: Aspect, i: Int, enumFacing: EnumFacing): Int = 0

  override def takeEssentia(aspect: Aspect, i: Int, enumFacing: EnumFacing): Int = 0

  override def setSuction(aspect: Aspect, i: Int): Unit = {}

  override def getSuctionType(enumFacing: EnumFacing): Aspect = null

  override def getSuctionAmount(enumFacing: EnumFacing): Int = 0

  override def getEssentiaAmount(enumFacing: EnumFacing): Int = 0

  override def getEssentiaType(enumFacing: EnumFacing): Aspect = null

  override def isConnectable(enumFacing: EnumFacing): Boolean = {
    val blockfacing:EnumFacing=worldObj.getBlockState(pos).getValue(BlockRepairer.FACING).asInstanceOf[EnumFacing]
    blockfacing==enumFacing
  }
//Essentia Stuff
  override def canOutputTo(enumFacing: EnumFacing): Boolean = true

  override def canInputFrom(enumFacing: EnumFacing): Boolean = true

  override def getMinimumSuction: Int = 0

  override def doesContainerAccept(aspect: Aspect): Boolean = false

  override def doesContainerContain(aspectList: AspectList): Boolean = false

  override def addToContainer(aspect: Aspect, i: Int): Int = 0

  override def containerContains(aspect: Aspect): Int = 0

  override def takeFromContainer(aspect: Aspect, i: Int): Boolean = false

  override def takeFromContainer(aspectList: AspectList): Boolean = false

  override def getAspects: AspectList = {
    val aspectList:AspectList=new AspectList(    )
    if (inventory!=null && inventory.isItemStackDamageable)
      {
        aspectList.add(Aspect.ENTROPY,inventory.getItemDamage)
      }
    return aspectList
  }

  override def setAspects(aspectList: AspectList): Unit = {}

  override def doesContainerContainAmount(aspect: Aspect, i: Int): Boolean = false
// Inventory stuff
  override def decrStackSize(index: Int, count: Int): ItemStack = {
  if (this.inventory!=null) {
    var itemStack:ItemStack=null
    if (this.inventory.stackSize <= count) {
      itemStack=this.inventory
      this.inventory=null
      this.markDirty()
      return itemStack
    }
    else {
      itemStack=this.inventory.splitStack(count)
      if (this.inventory.stackSize <= 0)
        this.inventory=null
      this.markDirty()
      return itemStack

    }
  }
  else
    return null
}


  override def markDirty(): Unit = {
    super.markDirty()
  worldObj.markBlockForUpdate(pos)
  }

  override def closeInventory(player: EntityPlayer): Unit = {}

  override def getSizeInventory: Int = 1

  override def getInventoryStackLimit: Int = 1

  override def clear(): Unit = {inventory=null}

  override def isItemValidForSlot(index: Int, stack: ItemStack): Boolean = true

  override def openInventory(player: EntityPlayer): Unit = {}

  override def getFieldCount: Int = 0

  override def getField(id: Int): Int = 0

  override def setInventorySlotContents(index: Int, stack: ItemStack): Unit = {
    inventory=stack
    this.markDirty()
  }

  override def isUseableByPlayer(player: EntityPlayer): Boolean = true

  override def getStackInSlot(index: Int): ItemStack = inventory

  override def setField(id: Int, value: Int): Unit = {}

  override def getDisplayName: IChatComponent = new ChatComponentText(getCommandSenderName)

 def getCommandSenderName: String = LibNames.REPAIRER

  override def hasCustomName: Boolean = false

  override def getSlotsForFace(side: EnumFacing): Array[Int] = Array[Int](0)

  override def canExtractItem(index: Int, stack: ItemStack, direction: EnumFacing): Boolean = true

  override def canInsertItem(index: Int, itemStackIn: ItemStack, direction: EnumFacing): Boolean = true

  override def removeStackFromSlot(index: Int): ItemStack =
  {
    if(index!=0)
      null
    else {
      val stack = inventory;
      inventory=null;
      stack
    }
  }

  override def getName: String = getCommandSenderName
}
